minoc/src/main/java/org/opencadc/minoc/MinocConfig.java

```diff
@@ -0,0 +1,386 @@
+            }
+        }
+
+        String schema = mvp.getFirstPropertyValue(SCHEMA_KEY);
+        sb.append("\n\t").append(SCHEMA_KEY).append(": ");
+        if (schema == null) {
+            sb.append("MISSING");
+            ok = false;
+        } else {
+            sb.append("OK");
+        }
+        
+        // optional
+        List<String> trusted = mvp.getProperty(TRUST_KEY);
+        if (trusted != null) {
+            for (String s : trusted) {
+                sb.append("\n\t").append(TRUST_KEY + "=").append(s);
+                try {
+                    URI uri = new URI(s);
+                    sb.append(" OK");
+                } catch (URISyntaxException ex) {
+                    sb.append(" INVALID");
+                    ok = false;
+                }
+            }
+        }
+        
+        // optional
+        List<String> readGrants = mvp.getProperty(READ_GRANTS_KEY);
+        if (readGrants != null) {
+            for (String s : readGrants) {
+                sb.append("\n\t").append(READ_GRANTS_KEY + "=").append(s);
+                try {
+                    URI u = new URI(s);
+                    sb.append(" OK");
+                } catch (URISyntaxException ex) {
+                    sb.append(" INVALID");
+                    ok = false;
+                }
+            }
+        }
+
+        // optional
+        List<String> writeGrants = mvp.getProperty(WRITE_GRANTS_KEY);
+        if (writeGrants != null) {
+            for (String s : writeGrants) {
+                sb.append("\n\t").append(WRITE_GRANTS_KEY + "=").append(s);
+                try {
+                    URI u = new URI(s);
+                    sb.append(" OK");
+                } catch (URISyntaxException ex) {
+                    sb.append(" INVALID");
+                    ok = false;
+                }
+            }
+        }
+        
+        // optional
+        List<String> rawRecNS = mvp.getProperty(RECOVERABLE_NS_KEY);
+        if (rawRecNS != null) {
+            for (String s : rawRecNS) {
+                sb.append("\n\t").append(RECOVERABLE_NS_KEY + "=").append(s);
+                try {
+                    Namespace ns = new Namespace(s);
+                } catch (Exception ex) {
+                    sb.append(" INVALID");
+                }
+            }
+        }
+
+        if (!ok) {
+            throw new IllegalStateException(sb.toString());
+        }
+    }
+    
+    public MultiValuedProperties getProperties() {
+        return configProperties;
+    }
+
+    public Map<URI, byte[]> getTrustedServices() {
+        syncKeys();
+        return trustedServices;
+    }
+    
+    private void syncKeys() {
+        if (!trustedServiceKeySync) {
+            int numFails = 0;
+            RegistryClient reg = new RegistryClient();
+            // check map for null keys and try to retrieve them
+            for (Map.Entry<URI,byte[]> me : trustedServices.entrySet()) {
+                if (me.getValue() == null) {
```

Comment:
I don't see how right now, the value is not `null` at this point. Can they be loaded from somewhere else?

Reply:
the syncKeys() can fail and leave the map entry with null values (the ctor only adds keys with null values)... for example if minoc starts up before a trusted service or fails to get the pub key

I'm not that happy with the rigidity of the check, especially because it gets called on every request (with a preauth key in the URL)... bad config can effect every such call so it isn't great right now.